scipy/stats/_mannwhitneyu.py

```diff
@@ -0,0 +1,430 @@
+import numpy as np
+from dataclasses import make_dataclass
+from collections import namedtuple
+from scipy import special
+from scipy import stats
+
+
+def _broadcast_concatenate(x, y, axis):
+    '''Broadcast then concatenate arrays, leaving concatenation axis last'''
+    x = np.moveaxis(x, axis, -1)
+    y = np.moveaxis(y, axis, -1)
+    z = np.broadcast(x[..., 0], y[..., 0])
+    x = np.broadcast_to(x, z.shape + (x.shape[-1],))
+    y = np.broadcast_to(y, z.shape + (y.shape[-1],))
+    z = np.concatenate((x, y), axis=-1)
+    return x, y, z
+
+
+class _MWU:
+    '''Distribution of MWU statistic under the null hypothesis'''
+    # Possible improvement: if m and n are small enough, use integer arithmetic
+
+    def __init__(self):
+        '''Minimal initializer'''
+        self._fmnks = -np.ones((1, 1, 1))
+
+    def pmf(self, k, m, n):
+        '''Probability mass function'''
+        self._resize_fmnks(m, n, np.max(k))
+        # could loop over just the unique elements, but probably not worth
+        # the time to find them
+        for i in np.ravel(k):
+            self._f(m, n, i)
+        return self._fmnks[m, n, k] / special.binom(m + n, m)
+
+    def cdf(self, k, m, n):
+        '''Cumulative distribution function'''
+        # We could use the fact that the distribution is symmetric to avoid
+        # summing more than m*n/2 terms, but it might not be worth the
+        # overhead. Let's leave that to an improvement.
+        pmfs = self.pmf(np.arange(0, np.max(k) + 1), m, n)
+        cdfs = np.cumsum(pmfs)
```

Comment:
If `k` is an integer, that is essentially what is happening because of the previous line. But it's vectorized, so if `k` is an array, we have to sum through the maximum value in `k`.

Reply:
ok, I got it